lib/ct.go

```diff
@@ -0,0 +1,108 @@
+package lib
+
+import (
+	"crypto/x509"
+	"encoding/asn1"
+	"encoding/base64"
+	"encoding/json"
+	"log"
+	"net/http"
+	"sync"
+	"time"
+
+	ctx509 "github.com/google/certificate-transparency-go/x509"
+	ctutil "github.com/google/certificate-transparency-go/x509util"
+)
+
+// https://github.com/google/certificate-transparency-community-site/blob/master/docs/google/known-logs.md
+const knownLogsAddr = "https://www.gstatic.com/ct/log_list/v2/log_list.json"
+
+var (
+	knownLogs     map[string]*ctLog
+	knownLogsOnce sync.Once
+)
+
+type ctLog struct {
+	operator string
+	url      string
+}
+
+func parseSCTList(cert *x509.Certificate) []*simpleSCT {
+	// ctutil contains a fork of crypto/x509 with support for SCTs. We must re-parse the
+	// whole certificate to get at them, so do a quick check to see if the SCT extension
+	// is present before re-parsing the cert unnecessarily.
+	if !hasSCTs(cert) {
+		return nil
+	}
+
+	var sctList []*simpleSCT
+	if scts, err := ctutil.ParseSCTsFromCertificate(cert.Raw); err == nil {
+		for _, sct := range scts {
+			id := sct.LogID.KeyID[:]
+			ssct := &simpleSCT{
+				Version:            uint64(sct.SCTVersion),
+				LogID:              id,
+				Timestamp:          time.UnixMilli(int64(sct.Timestamp)),
+				SignatureAlgorithm: sctSignatureAlg(sct.Signature.Algorithm),
+			}
+			if log := getLogByID(id); log != nil {
+				ssct.LogOperator = log.operator
+				ssct.LogURL = log.url
+			}
+			sctList = append(sctList, ssct)
+		}
+	}
+	return sctList
+}
+
+func hasSCTs(cert *x509.Certificate) bool {
+	for _, e := range cert.Extensions {
+		if e.Id.Equal(asn1.ObjectIdentifier(ctx509.OIDExtensionCTSCT)) {
+			return true
+		}
+	}
+	return false
+}
+
+func getLogByID(id []byte) *ctLog {
+	knownLogsOnce.Do(func() {
+		client := &http.Client{
+			// Set a timeout so we don't block forever on broken servers.
+			Timeout: 5 * time.Second,
+		}
+
+		resp, err := client.Get(knownLogsAddr)
```

Comment:
Making a network call to get the current list of known logs during, say, `certigo dump` is less than ideal (but will only happen if the cert has SCTs). Another alternative is to hardcode a list of known logs, but I'm not sure we should sign up to maintain that. And finally, a third option would be to only fetch the log URL and operator if an explicit flag was passed, but that would require a fair bit of plumbing to accomplish (but is probably the right thing to do?)

Reply:
I decided to keep a hardcoded list to avoid the network call at runtime.